File: Unique Binary Search Trees.py
class Solution:
    def numTrees(self, n: int) -> int:
        dp = {0: 1, 1: 1, 2: 2}
        def recur(r):
            if r in dp:
                return dp[r]

            # Choose mid value
            summ = 0
            for i in range(r):
                summ += recur(i) * recur(r-1-i)
            dp[r] = summ
            return summ
        return recur(n)

File: Unique Binary Search Trees.go
var dp = map[int]int{0:1, 1:1, 2:2}

func recur(r int) int {
    if v, ok := dp[r]; ok {
        return v
    }

    summ := 0;
    for i := 0; i < r; i++ {
        summ += recur(i) * recur(r-i-1)
    }
    dp[r] = summ
    return summ
}

func numTrees(n int) int {
    return recur(n)
}
